aspnetcore/blazor/samples/5.0/BlazorSample_Server/Pages/file-uploads/FileUpload3.razor

```diff
@@ -0,0 +1,111 @@
+@page "/file-upload-3"
+@using System 
+@using System.IO
+@using Microsoft.AspNetCore.Hosting
+@using Microsoft.Extensions.Logging
+@inject ILogger<FileUpload3> Logger
+@inject IWebHostEnvironment Environment
+
+<h3>Upload Files</h3>
+
+<p>
+    <label>
+        Max file size:
+        <input type="number" @bind="maxFileSize" />
+    </label>
+</p>
+
+<p>
+    <label>
+        Max allowed files:
+        <input type="number" @bind="maxAllowedFiles" />
+    </label>
+</p>
+
+<p>
+    <label>
+        Upload up to @maxAllowedFiles of up to @maxFileSize bytes:
+        <InputFile OnChange="@LoadFiles" multiple />
+    </label>
+</p>
+
+@if (isLoading)
+{
+    <p>Progress: @string.Format("{0:P0}", progressPercent)</p>
+}
+else
+{
+    <ul>
+        @foreach (var file in loadedFiles)
+        {
+            <li>
+                <ul>
+                    <li>Name: @file.Name</li>
+                    <li>Last modified: @file.LastModified.ToString()</li>
+                    <li>Size (bytes): @file.Size</li>
+                    <li>Content type: @file.ContentType</li>
+                </ul>
+            </li>
+        }
+    </ul>
+}
+
+@code {
+    private List<IBrowserFile> loadedFiles = new();
+    private long maxFileSize = 1024 * 1024 * 16;
+    private int maxAllowedFiles = 3;
+    private bool isLoading;
+    private decimal progressPercent = 0.0M;
+
+    private async Task LoadFiles(InputFileChangeEventArgs e)
+    {
+        isLoading = true;
+        loadedFiles.Clear();
+
+        foreach (var file in e.GetMultipleFiles(maxAllowedFiles))
+        {
+            try
+            {
+                var trustedFileName = Path.GetRandomFileName();
+                var path = Path.Combine(Environment.ContentRootPath, 
+                    Environment.EnvironmentName, "unsafe_uploads", trustedFileName);
+
+                await using FileStream writeStream = new(path, FileMode.Create);
+                using var readStream = file.OpenReadStream(maxFileSize);
+
+                var bytes = new byte[readStream.Length];
```

Comment:
Hmm this may be problematic. Presumably file upload progress is desired for larger files. This would force large amounts of server-side memory allocations for the large files (which could pose a security concern). I think it may be better to limit this to a small buffer (ex. 30kb), and then write the buffer to the filestream in the `while` loop before requesting the next chunk.